### PR TITLE
Adding iteration to ProgressBar to make it a more useful utility

### PR DESCRIPTION
**Pull request by henryiii**
As discussed in #10812, this simply adds the ability to iterate over a ProgessBar (no other API changes). Added a test and a mention in the relevant example.

### CODE REVIEW

File: IPython/core/display.py [hunk 1]
```diff
@@ -774,6 +774,19 @@ def progress(self, value):
...
+    def __iter__(self):
+        self.display()
+        self._progress = -1 # First iteration is 0
+        return self
+
+    def __next__(self):
+        """Returns current value and increments display by one."""
+        self.progress += 1
+        if self.progress < self.total:
+            return self.progress
```
Comment (mariusvniekerk) [inline IPython/core/display.py]: Increment afterwards, that way you don't need to initialize to - 1.

Comment (henryiii) [reply IPython/core/display.py]: This is just a function, not a generator, so anything after the return will not run. You have to either have a -1 initial value (which is never printed), or you have to add a new variable to tell the difference between before-iteration 0 and iteration 0. You also really want the final iteration to be 4/5, to match the range, and to provide a visual indication that it's finished by changing to 5/5.

File: IPython/core/tests/test_display.py [hunk 2]
```diff
@@ -196,11 +196,18 @@ def test_displayobject_repr():
...
     p.progress = 5
     nt.assert_equal(p._repr_html_(), "<progress style='width:100%' max='10' value='5'></progress>")
 
+def test_progress_iter(capsys):
+    for i in display.ProgressBar(5):
+        out, err = capsys.readouterr()
+        nt.assert_in('{0}/5'.format(i), out)
+    out, err = capsys.readouterr()
+    nt.assert_in('5/5', out)
+
```
Comment (henryiii) [inline IPython/core/tests/test_display.py]: Oops, I guess this is nose running nose tests, not pytest running nose tests.

The nose project hasn't receive a commit in about a year, I think pytest would be a good idea :) I'll fix this.

**Comment by henryiii** [comment]
Is there a way to test in a normal terminal environment? It seems that the tests run in a environment where display is not supported.

**Comment by takluyver** [comment]
I don't think it's the environment that's involved there: it looks like `CapturingDisplayPublisher` (which is used by `capture_output`) hasn't been updated with the `transient=` keyword argument that the regular `DisplayPublisher` has. I guess you'll need to update that, or use `capture_output(display=False)` and let it capture the output on stdout.

**Comment by henryiii** [comment]
I've used the `display=False` to make sure this works for IPython 5 or 6, and separately patched the display issue for IPython 5 and 6.

**Comment by takluyver** [comment]
@mariusvniekerk @Carreau are you happy with this?

**Review by mariusvniekerk — APPROVED** [review]
Looks good

**Comment by Carreau** [comment]
+1, I'm gessing backport to have identical functionality on Python2.

**Comment by henryiii** [comment]
Yes, it should be in both.

**Pull request merged**